File: frontend/src/pages/UserManagement.js
import React, { useState, useEffect } from 'react';
import { useAuth } from '../hooks/useAuth';
import { endpoints } from '../services/api';
import toast from 'react-hot-toast';

export default function UserManagement() {
  const { user } = useAuth();
  const [users, setUsers] = useState([]);
  const [loading, setLoading] = useState(true);
  const [stats, setStats] = useState(null);
  const [promoteUser, setPromoteUser] = useState('');
  const [promoteRole, setPromoteRole] = useState('admin');

  useEffect(() => {
    if (user?.role === 'admin') {
      fetchUsers();
      fetchStats();
    }
  }, [user]);

  const fetchUsers = async () => {
    try {
      const response = await endpoints.getUsers();
      setUsers(response.data);
    } catch (error) {
      toast.error('Failed to fetch users');
      console.error(error);
    } finally {
      setLoading(false);
    }
  };

  const fetchStats = async () => {
    try {
      const response = await endpoints.getUserStats();
      setStats(response.data);
    } catch (error) {
      console.error('Failed to fetch user stats:', error);
    }
  };

  const updateUserRole = async (userId, newRole) => {
    try {
      await endpoints.updateUserRole(userId, newRole);
      toast.success(`User role updated to ${newRole}`);
      fetchUsers();
      fetchStats();
    } catch (error) {
      toast.error(error.response?.data?.error || 'Failed to update user role');
    }
  };

  const updateUserStatus = async (userId, isActive) => {
    try {
      await endpoints.updateUserStatus(userId, isActive);
      toast.success(`User ${isActive ? 'activated' : 'deactivated'}`);
      fetchUsers();
      fetchStats();
    } catch (error) {
      toast.error(error.response?.data?.error || 'Failed to update user status');
    }
  };

  const deleteUser = async (userId) => {
    if (!window.confirm('Are you sure you want to deactivate this user?')) return;
    
    try {
      await endpoints.deleteUser(userId);
      toast.success('User deactivated');
      fetchUsers();
      fetchStats();
    } catch (error) {
      toast.error(error.response?.data?.error || 'Failed to delete user');
    }
  };

  const handlePromoteUser = async (e) => {
    e.preventDefault();
    if (!promoteUser.trim()) {
      toast.error('Please enter a GitHub username');
      return;
    }

    try {
      await endpoints.promoteGitHubUser(promoteUser, promoteRole);
      toast.success(`GitHub user "${promoteUser}" promoted to ${promoteRole}`);
      setPromoteUser('');
      fetchUsers();
      fetchStats();
    } catch (error) {
      toast.error(error.response?.data?.error || 'Failed to promote user');
    }
  };

  if (user?.role !== 'admin') {
    return (
      <div className="text-center py-16">
        <h2 className="text-2xl font-bold text-gray-900 mb-4">Access Denied</h2>
        <p className="text-gray-600">You need admin privileges to access user management.</p>
      </div>
    );
  }

  if (loading) {
    return (
      <div className="flex items-center justify-center py-16">
        <div className="animate-spin rounded-full h-8 w-8 border-b-2 border-blue-600"></div>
      </div>
    );
  }

  return (
    <div className="max-w-7xl mx-auto px-4 py-8">
      <div className="mb-8">
        <h1 className="text-3xl font-bold text-gray-900 mb-2">User Management</h1>
        <p className="text-gray-600">Manage user roles and permissions</p>
      </div>

      {/* Statistics Cards */}
      {stats && (
        <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-6 mb-8">
          <div className="bg-white p-6 rounded-lg shadow-md">
            <h3 className="text-sm font-medium text-gray-500 uppercase tracking-wide">Total Users</h3>
            <p className="text-3xl font-bold text-blue-600">{stats.total_users}</p>
          </div>
          <div className="bg-white p-6 rounded-lg shadow-md">
            <h3 className="text-sm font-medium text-gray-500 uppercase tracking-wide">Active Users</h3>
            <p className="text-3xl font-bold text-green-600">{stats.active_users}</p>
          </div>
          <div className="bg-white p-6 rounded-lg shadow-md">
            <h3 className="text-sm font-medium text-gray-500 uppercase tracking-wide">Admins</h3>
            <p className="text-3xl font-bold text-purple-600">{stats.role_distribution.admin}</p>
          </div>
          <div className="bg-white p-6 rounded-lg shadow-md">
            <h3 className="text-sm font-medium text-gray-500 uppercase tracking-wide">Contributors</h3>
            <p className="text-3xl font-bold text-orange-600">{stats.role_distribution.contributor}</p>
          </div>
        </div>
      )}

      {/* Promote User Form */}
      <div className="bg-white p-6 rounded-lg shadow-md mb-8">
        <h2 className="text-xl font-semibold mb-4">Promote GitHub User</h2>
        <form onSubmit={handlePromoteUser} className="flex gap-4">
          <input
            type="text"
            placeholder="GitHub username"
            value={promoteUser}
            onChange={(e) => setPromoteUser(e.target.value)}
            className="flex-1 px-3 py-2 border border-gray-300 rounded-md focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          />
          <select
            value={promoteRole}
            onChange={(e) => setPromoteRole(e.target.value)}
            className="px-3 py-2 border border-gray-300 rounded-md focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          >
            <option value="admin">Admin</option>
            <option value="contributor">Contributor</option>
            <option value="user">User</option>
          </select>
          <button
            type="submit"
            className="px-6 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:ring-2 focus:ring-blue-500"
          >
            Promote
          </button>
        </form>
        <p className="text-sm text-gray-500 mt-2">
          User must have logged in at least once via GitHub OAuth before promotion.
        </p>
      </div>

      {/* Users Table */}
      <div className="bg-white rounded-lg shadow-md overflow-hidden">
        <div className="px-6 py-4 border-b border-gray-200">
          <h2 className="text-xl font-semibold">All Users ({users.length})</h2>
        </div>
        
        <div className="overflow-x-auto">
          <table className="w-full">
            <thead className="bg-gray-50">
              <tr>
                <th className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                  User
                </th>
                <th className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                  Email
                </th>
                <th className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                  Role
                </th>
                <th className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                  Status
                </th>
                <th className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                  Last Login
                </th>
                <th className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                  Actions
                </th>
              </tr>
            </thead>
            <tbody className="bg-white divide-y divide-gray-200">
              {users.map((u) => (
                <tr key={u.id} className={!u.is_active ? 'bg-gray-50 opacity-75' : ''}>
                  <td className="px-6 py-4 whitespace-nowrap">
                    <div className="flex items-center">
                      {u.github_avatar_url && (
                        <img 
                          src={u.github_avatar_url} 
                          alt={u.display_name} 
                          className="h-8 w-8 rounded-full mr-3"
                        />
                      )}
                      <div>
                        <div className="text-sm font-medium text-gray-900">
                          {u.display_name || u.username}
                        </div>
                        <div className="text-sm text-gray-500">@{u.github_username}</div>
                      </div>
                    </div>
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-900">
                    {u.email}
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap">
                    <select
                      value={u.role}
                      onChange={(e) => updateUserRole(u.id, e.target.value)}
                      className="text-sm border-gray-300 rounded-md focus:ring-blue-500 focus:border-blue-500"
                      disabled={u.id === user.id} // Prevent self-role change
                    >
                      <option value="user">User</option>
                      <option value="contributor">Contributor</option>
                      <option value="admin">Admin</option>
                    </select>
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap">
                    <button
                      onClick={() => updateUserStatus(u.id, !u.is_active)}
                      className={`inline-flex px-2 py-1 text-xs font-semibold rounded-full ${
                        u.is_active
                          ? 'bg-green-100 text-green-800 hover:bg-green-200'
                          : 'bg-red-100 text-red-800 hover:bg-red-200'
                      }`}
                      disabled={u.id === user.id} // Prevent self-deactivation
                    >
                      {u.is_active ? 'Active' : 'Inactive'}
                    </button>
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">
                    {u.last_login ? new Date(u.last_login).toLocaleDateString() : 'Never'}
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap text-sm font-medium">
                    {u.id !== user.id && (
                      <button
                        onClick={() => deleteUser(u.id)}
                        className="text-red-600 hover:text-red-900"
                      >
                        Deactivate
                      </button>
                    )}
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>

        {users.length === 0 && (
          <div className="text-center py-16">
            <p className="text-gray-500">No users found</p>
          </div>
        )}
      </div>

      {/* Recent Logins */}
      {stats?.recent_logins && stats.recent_logins.length > 0 && (
        <div className="mt-8 bg-white rounded-lg shadow-md overflow-hidden">
          <div className="px-6 py-4 border-b border-gray-200">
            <h2 className="text-xl font-semibold">Recent Logins</h2>
          </div>
          <div className="p-6">
            <div className="space-y-3">
              {stats.recent_logins.map((login, index) => (
                <div key={index} className="flex justify-between items-center py-2 border-b border-gray-100">
                  <div>
                    <span className="font-medium">{login.display_name || login.github_username}</span>
                    <span className={`ml-2 px-2 py-1 text-xs rounded-full ${
                      login.role === 'admin' ? 'bg-purple-100 text-purple-800' :
                      login.role === 'contributor' ? 'bg-orange-100 text-orange-800' :
                      'bg-gray-100 text-gray-800'
                    }`}>
                      {login.role}
                    </span>
                  </div>
                  <span className="text-sm text-gray-500">
                    {login.last_login ? new Date(login.last_login).toLocaleString() : 'Never'}
                  </span>
                </div>
              ))}
            </div>
          </div>
        </div>
      )}
    </div>
  );
}

File: backend/app/routes/user_management.py
from flask import Blueprint, request, jsonify
from app.models.user import User
from app.services.auth import require_admin, get_current_user
from app import db
from datetime import datetime

bp = Blueprint('user_management', __name__, url_prefix='/admin/users')

@bp.route('', methods=['GET'])
@require_admin
def list_users():
    """Get all users for admin management"""
    try:
        users = User.query.all()
        user_list = []
        for user in users:
            user_data = user.to_dict()
            # Add additional info
            user_data['login_count'] = 1 if user.last_login else 0
            user_list.append(user_data)
        
        return jsonify(user_list), 200
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@bp.route('/<user_id>', methods=['GET'])
@require_admin  
def get_user(user_id):
    """Get specific user details"""
    try:
        user = User.query.get(user_id)
        if not user:
            return jsonify({'error': 'User not found'}), 404
        
        return jsonify(user.to_dict()), 200
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@bp.route('/<user_id>/role', methods=['PUT'])
@require_admin
def update_user_role(user_id):
    """Update user role"""
    try:
        user = User.query.get(user_id)
        if not user:
            return jsonify({'error': 'User not found'}), 404
        
        data = request.json
        new_role = data.get('role')
        
        if new_role not in ['admin', 'contributor', 'user']:
            return jsonify({'error': 'Invalid role. Must be admin, contributor, or user'}), 400
        
        # Prevent removing the last admin
        if user.role == 'admin' and new_role != 'admin':
            admin_count = User.query.filter_by(role='admin').count()
            if admin_count <= 1:
                return jsonify({'error': 'Cannot remove the last admin user'}), 400
        
        user.role = new_role
        user.updated_at = datetime.utcnow()
        db.session.commit()
        
        return jsonify({
            'message': f'User role updated to {new_role}',
            'user': user.to_dict()
        }), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@bp.route('/<user_id>/status', methods=['PUT'])
@require_admin
def update_user_status(user_id):
    """Activate or deactivate user"""
    try:
        user = User.query.get(user_id)
        if not user:
            return jsonify({'error': 'User not found'}), 404
        
        data = request.json
        is_active = data.get('is_active')
        
        if is_active is None:
            return jsonify({'error': 'is_active field is required'}), 400
        
        # Prevent deactivating the last admin
        current_admin = get_current_user()
        if user.role == 'admin' and not is_active:
            admin_count = User.query.filter_by(role='admin', is_active=True).count()
            if admin_count <= 1:
                return jsonify({'error': 'Cannot deactivate the last admin user'}), 400
        
        user.is_active = bool(is_active)
        user.updated_at = datetime.utcnow()
        db.session.commit()
        
        status = 'activated' if is_active else 'deactivated'
        return jsonify({
            'message': f'User {status} successfully',
            'user': user.to_dict()
        }), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@bp.route('/<user_id>', methods=['DELETE'])
@require_admin
def delete_user(user_id):
    """Delete user (soft delete by deactivating)"""
    try:
        user = User.query.get(user_id)
        if not user:
            return jsonify({'error': 'User not found'}), 404
        
        # Prevent deleting the last admin
        if user.role == 'admin':
            admin_count = User.query.filter_by(role='admin', is_active=True).count()
            if admin_count <= 1:
                return jsonify({'error': 'Cannot delete the last admin user'}), 400
        
        # Prevent self-deletion
        current_admin = get_current_user()
        if current_admin and current_admin.id == user.id:
            return jsonify({'error': 'Cannot delete your own account'}), 400
        
        # Soft delete - just deactivate
        user.is_active = False
        user.updated_at = datetime.utcnow()
        db.session.commit()
        
        return jsonify({'message': 'User deactivated successfully'}), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@bp.route('/promote/<github_username>', methods=['POST'])
@require_admin
def promote_github_user(github_username):
    """Promote a GitHub user to admin by username"""
    try:
        user = User.query.filter_by(github_username=github_username).first()
        if not user:
            return jsonify({'error': f'User with GitHub username "{github_username}" not found. User must login first.'}), 404
        
        data = request.json
        new_role = data.get('role', 'admin')
        
        if new_role not in ['admin', 'contributor', 'user']:
            return jsonify({'error': 'Invalid role. Must be admin, contributor, or user'}), 400
        
        user.role = new_role
        user.updated_at = datetime.utcnow()
        db.session.commit()
        
        return jsonify({
            'message': f'GitHub user "{github_username}" promoted to {new_role}',
            'user': user.to_dict()
        }), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@bp.route('/stats', methods=['GET'])
@require_admin
def get_user_stats():
    """Get user statistics"""
    try:
        total_users = User.query.count()
        active_users = User.query.filter_by(is_active=True).count()
        admin_count = User.query.filter_by(role='admin', is_active=True).count()
        contributor_count = User.query.filter_by(role='contributor', is_active=True).count()
        user_count = User.query.filter_by(role='user', is_active=True).count()
        
        recent_logins = User.query.filter(User.last_login.isnot(None))\
                                 .order_by(User.last_login.desc())\
                                 .limit(10).all()
        
        return jsonify({
            'total_users': total_users,
            'active_users': active_users,
            'role_distribution': {
                'admin': admin_count,
                'contributor': contributor_count,
                'user': user_count
            },
            'recent_logins': [
                {
                    'github_username': user.github_username,
                    'display_name': user.display_name,
                    'last_login': user.last_login.isoformat() if user.last_login else None,
                    'role': user.role
                }
                for user in recent_logins
            ]
        }), 200
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500
